tests/legacy/test_read_exceptions.py

```diff
@@ -203,3 +206,80 @@ def test_timeout_stream(self, generate_task, real_x_series_device):
+
+        samples_to_read = 75
+        clocks_to_give = 100
+        sample_rate = 1000
+        # Choose a sentinel value that can't be returned by the read.
+        sentinel_value = -100.0
+        number_of_channels = 2
+
+        read_task = generate_task()
+        sample_clk_task = generate_task()
+        # Use a counter output pulse train task as the sample clock source
+        # for both the AI and AO tasks.
+        sample_clk_task.co_channels.add_co_pulse_chan_freq(
+            f"{real_x_series_device.name}/ctr0", freq=sample_rate, idle_state=Level.LOW
+        )
+        sample_clk_task.timing.cfg_implicit_timing(
+            samps_per_chan=clocks_to_give, sample_mode=AcquisitionType.FINITE
+        )
+        sample_clk_task.control(TaskMode.TASK_COMMIT)
+
+        samp_clk_terminal = f"/{real_x_series_device.name}/Ctr0InternalOutput"
+
+        channel_names = f"{real_x_series_device.ai_physical_chans[0].name},{real_x_series_device.ai_physical_chans[1].name}"
+        read_task.ai_channels.add_ai_voltage_chan(channel_names, max_val=10, min_val=-10)
+        read_task.timing.cfg_samp_clk_timing(
+            sample_rate, source=samp_clk_terminal, sample_mode=AcquisitionType.CONTINUOUS
+        )
+
+        reader = nidaqmx.stream_readers.AnalogMultiChannelReader(read_task.in_stream)
+
+        # Start the read task before the clock.
+        read_task.start()
+
+        # Generate all the clocks.
+        sample_clk_task.start()
+        sample_clk_task.wait_until_done()
+        sample_clk_task.stop()
+
+        # Set read timeout.
+        read_task.in_stream.timeout = 2
+
+        # Do a partial read which should succeed.
+        data = numpy.full(
+            (number_of_channels, samples_to_read), sentinel_value, dtype=numpy.float64
+        )
+        num_values_read = reader.read_many_sample(
+            data, number_of_samples_per_channel=samples_to_read, timeout=2.0
+        )
+        assert num_values_read == samples_to_read
+        # All the data should have been overwritten.
+        assert (data != sentinel_value).any()
+
+        # Now read more data than is available.
+        data = numpy.full(
+            (number_of_channels, samples_to_read), sentinel_value, dtype=numpy.float64
+        )
+        with pytest.raises(nidaqmx.DaqReadError) as timeout_exception:
+            num_values_read = reader.read_many_sample(
+                data, number_of_samples_per_channel=samples_to_read, timeout=2.0
+            )
+
+        assert timeout_exception.value.error_code == DAQmxErrors.SAMPLES_NOT_YET_AVAILABLE
+        # We should have read a partial dataset.
+        number_of_samples_expected = clocks_to_give - samples_to_read
+        assert timeout_exception.value.samps_per_chan_read == number_of_samples_expected
+        # The data that wasn't read should have been unmodified.
+        assert sum(numpy.count_nonzero(element == sentinel_value) for element in data) == 75
```

Comment:
The number of samples per channel is 75 and based on  the timeout that we have given it should read 50 samples and 25 samples should be unmodified in each channel. But what I actually observed is in one channel 50 samples was read and in another channel only 25 samples was read. I wasn't able to figure out the exact reason for this discrepancy. Do you have any suggestions on the above behavior @bkeryan?

Reply:
It's using the wrong array as the source of the assignment.
```
        if numpy_array.ndim > 1:
            (number_of_channels, number_of_samples_per_channel) = numpy_array.shape	
            read_array = numpy.array(grpc_array).reshape((number_of_channels, samps_per_chan_read))	
            numpy_array[:, :samps_per_chan_read].flat[:grpc_array_size] = grpc_array
```

Once you fix that, it will fail for a different reason.

The server should return something like this (but with 10x the data)
```
A0 A1 A2 X X X X B0 B1 B2 X X X X
```
```
>>> import numpy
>>> a = numpy.array([1,2,3,0,0,0,0,4,5,6,0,0,0,0])
>>> a
array([1, 2, 3, 0, 0, 0, 0, 4, 5, 6, 0, 0, 0, 0])
```

samps_per_chan_read should be 3, but we can't reshape a 1x14 array into 2x3:
```
>>> b = a.reshape((2,3))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot reshape array of size 14 into shape (2,3)
```

We can reshape it into 2x7:
```
>>> b = a.reshape((2,7))
>>> b
array([[1, 2, 3, 0, 0, 0, 0],
       [4, 5, 6, 0, 0, 0, 0]])
```

We can't assign a 2x3 slice based on an 2x7 array:
```
>>> c = numpy.full((2,7),-1)
>>> c
array([[-1, -1, -1, -1, -1, -1, -1],
       [-1, -1, -1, -1, -1, -1, -1]])
>>> c[:,:3] = b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not broadcast input array from shape (2,7) into shape (2,3)
```

But we can assign a 2x3 slice based on the 2x3 slice of a 2x7 array:
```
>>> c[:,:3] = b[:,:3]
>>> c
array([[ 1,  2,  3, -1, -1, -1, -1],
       [ 4,  5,  6, -1, -1, -1, -1]])
```

I think you need something like 
```
        if numpy_array.ndim > 1:
            (number_of_channels, number_of_samples_per_channel) = numpy_array.shape	
            read_array = numpy.array(grpc_array).reshape((number_of_channels, grpc_array_size // number_of_channels))	
            numpy_array[:, :samps_per_chan_read] = read_array[:, :samps_per_chan_read]
```

I also think that reducing the number of samples and comparing against an expected array like `numpy.array([[1, 2, 3, -1, -1, -1, -1], [4, 5, 6, -1, -1, -1, -1]])` will make the assertions simpler and more accurate.